### PR TITLE
Make update planner independent of parameters

### PR DESCRIPTION
**Pull request by mfussenegger**
This changes the logic for update statements in the planner and executor
to work with the unbound analyzer result.

This is similar to the work that's been done in
https://github.com/crate/crate/pull/6561

This'll enable us to add scalar subquery support for UPDATE statements
later on.

(This now also includes the changes of https://github.com/crate/crate/pull/6577 which are more or less eliminated again. I think I could also remove those commits altogether)

### CODE REVIEW

File: sql/src/main/java/io/crate/executor/transport/task/DeleteByIdTask.java [hunk 1]
```diff
@@ -91,32 +90,29 @@ public void execute(final RowConsumer consumer, Row parameters) {
...
         for (int i = 0; i < response.itemIndices().size(); i++) {
             ShardResponse.Failure failure = response.failures().get(i);
             if (failure == null) {
                 rowCount[0] += 1;
-            } else if (!failure.versionConflict() && !failure.message().contains("Document not found")) {
+            } else if (!failure.versionConflict() && !(failure.message().contains("Document not found") || failure.message().contains("document missing"))) {
                 throw new RuntimeException(failure.message());
             }
         }
+        Exception failure = response.failure();
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/task/DeleteByIdTask.java]: shouldn't we do that before checking each items failure?

File: sql/src/main/java/io/crate/executor/transport/task/UpdateByIdTask.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    @Override
+    public List<CompletableFuture<Long>> executeBulk(List<Row> bulkParams) {
+        HashMap<ShardId, ShardUpsertRequest> requestsByShard = new HashMap<>();
+        ShardUpsertRequest.Builder requestBuilder = createBuilder.apply(true);
+        int location = 0;
+        // idx=location; value=resultIdx
+        IntArrayList bulkIndices = new IntArrayList(bulkParams.size() * updateById.docKeys().size());
+        for (int resultIdx = 0; resultIdx < bulkParams.size(); resultIdx++) {
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/task/UpdateByIdTask.java]: would it makes sense to extract this into a shared method as this looks identically to whats happening at the DeleteByIdTask?

File: sql/src/main/java/io/crate/executor/transport/task/UpdateByIdTask.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        }
+        BulkShardResponseListener listener =
+            new BulkShardResponseListener(requestsByShard.size(), bulkParams.size(), bulkIndices);
+        for (ShardUpsertRequest request : requestsByShard.values()) {
+            shardUpsertAction.execute(request, listener);
+        }
+        return listener.rowCountFutures();
+    }
+
+    private int addRequests(int location, Row parameters, HashMap<ShardId, ShardUpsertRequest> requestsByShard, ShardUpsertRequest.Builder requestBuilder) {
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/task/UpdateByIdTask.java]: could we maybe share this logic with DeleteByIdTask as well? also is it require to expect a HashMap here?

File: sql/src/main/java/io/crate/planner/node/dml/LegacyUpsertById.java [hunk 4]
```diff
@@ -32,7 +32,12 @@
 import java.util.List;
 import java.util.UUID;
 
-public class UpsertById extends UnnestablePlan {
+/**
+ * @deprecated This should be replaced with a variant that doesn't depend on parameter values.
```
Comment (seut) [inline sql/src/main/java/io/crate/planner/node/dml/LegacyUpsertById.java]: maybe add a note that this is currently just only used for inserts

File: sql/src/main/java/io/crate/executor/transport/TransportExecutor.java [hunk 5]
```diff
@@ -307,8 +311,14 @@ public Task visitDeleteById(DeleteById plan, PlannerContext context) {
         }
 
         @Override
-        public Task visitUpsertById(UpsertById plan, PlannerContext context) {
-            return new UpsertByIdTask(
+        public Task visitUpdateById(UpdateById updateById, PlannerContext context) {
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql/src/main/java/io/crate/executor/transport/TransportExecutor.java]: Yes in the non-bulk case

**Pull request merged**